File: Days 31 - 40/Day 37/CreatePowerSet.cpp
#include <cmath>
#include <iostream>
#include <vector>

template <typename T>
std::vector<std::vector<T>> CreatePowerSet(const std::vector<T>& set) noexcept
{
	unsigned int subsets = static_cast<unsigned int>(std::pow(2, set.size()));

	std::vector<std::vector<T>> powerSet;
	powerSet.reserve(subsets);

	for (unsigned int i = 0; i < subsets; i++)
	{
		std::vector<T> currentSubset;
		currentSubset.reserve(set.size());

		for (unsigned int j = 0; j < subsets; j++)
		{
			if (int mask = 1 << j; (i & mask) == mask)
			{
				currentSubset.push_back(set[j]);
			}
		}

		powerSet.push_back(currentSubset);
	}

	return powerSet;
}

template <typename T>
std::ostream& operator <<(std::ostream& os, const std::vector<T>& set)
{
	os << "{";

	for (unsigned int i = 0; i < set.size(); i++)
	{
		os << " " << set[i];

		if (i != set.size() - 1)
		{
			os << ",";
		}
	}

	os << " }";

	return os;
}

int main(int argc, const char* argv[])
{
	std::vector<int> values { 1, 2, 3 };
	auto powerSet = CreatePowerSet(values);

	std::cout << powerSet << "\n";

	std::cin.get();

	return 0;
}

File: Days 31 - 40/Day 37/CreatePowerSet.cs
using System;
using System.Collections.Generic;

namespace DailyCodingProblem
{
	internal class Day37
	{
		private static int Main(string[] args)
		{
			List<int> values = new List<int> { 1, 2, 3, 4 };
			List<List<int>> powerSet = CreatePowerSet(values);

			PrintPowerSet(powerSet);

			Console.ReadLine();

			return 0;
		}

		private static List<List<T>> CreatePowerSet<T>(List<T> set)
		{
			int subsets = (int)Math.Pow(2, set.Count);
			List<List<T>> powerSet = new List<List<T>>(subsets);

			for (int i = 0; i < subsets; i++)
			{
				List<T> currentSubset = new List<T>(set.Count);

				for (int j = 0; j < subsets; j++)
				{
					int mask = 1 << j;

					if ((i & mask) == mask)
					{
						currentSubset.Add(set[j]);
					}
				}

				powerSet.Add(currentSubset);
			}

			return powerSet;
		}

		private static void PrintPowerSet<T>(List<List<T>> powerSet)
		{
			foreach (List<T> set in powerSet)
			{
				foreach (T value in set)
				{
					Console.Write($"{value} ");
				}

				Console.WriteLine();
			}
		}
	}
}